src/test-explorer/test-run-handler.ts

```diff
@@ -0,0 +1,171 @@
+import { ServerCommands } from "metals-languageclient";
+import * as vscode from "vscode";
+import {
+  CancellationToken,
+  commands,
+  TestController,
+  TestRunRequest,
+} from "vscode";
+import { debugServerFromUri, DebugSession } from "../scalaDebugger";
+import { analyzeTestRun } from "./analyze-test-run";
+import { testCache } from "./test-cache";
+import { DapEvent, TargetUri, TestItemMetadata } from "./types";
+
+// this id is used to mark DAP sessions created by TestController
+// thanks to that debug tracker knows which requests it should track and gather results
+export const testRunnerId = "scala-dap-test-runner";
+
+/**
+ * Register tracker which tracks all DAP sessions which are started with @constant {testRunnerId} kind.
+ * Dap sends execution result for every suite included in TestRun in a special event of kind 'testResult'.
+ * Tracker has to capture these events and store them all in map under debug session id as a key.
+ */
+vscode.debug.registerDebugAdapterTrackerFactory("scala", {
+  createDebugAdapterTracker(session) {
+    if (session.configuration.kind === testRunnerId) {
+      return {
+        onWillStartSession: () => testCache.setEmptySuiteResultsFor(session.id),
```

Comment:
Can we run multiple tests at the same time?

Reply:
Actually, could we not allow it? Sorry for not wording it clearer. Is there anything that could break on multiple sessions?